realm/src/main/java/io/realm/RealmQuery.java

```diff
@@ -49,23 +51,73 @@
  */
 public class RealmQuery<E extends RealmObject> {
 
-    private final Realm realm;
-    private final Table table;
-    private final LinkView view;
-    private final TableQuery query;
-    private final Map<String, Long> columns;
-    private final Class<E> clazz;
+    private BaseRealm realm;
+    private Class<E> clazz;
+    private String className;
+    private Map<String, Long> columns = new HashMap<String, Long>();
 
-    private static final String LINK_NOT_SUPPORTED_METHOD = "'%s' is not supported for link queries";
+    private Table table;
+    private LinkView view;
+    private TableQuery query;
 
     /**
-     * Creating a RealmQuery instance.
+     * Creates a query for objects of a given class from a {@link Realm}.
      *
      * @param realm  The realm to query within.
      * @param clazz  The class to query.
-     * @throws java.lang.RuntimeException Any other error.
+     * @return {@link RealmQuery} object. After building the query call one of the {@code find*} methods
+     * to run it.
+     */
+    public static <E extends RealmObject> RealmQuery<E> createQuery(Realm realm, Class<E> clazz) {
+        return new RealmQuery<E>(realm, clazz);
+    }
+
+    /**
+     * Creates a query for dynamic objects of a given type from a {@link DynamicRealm}.
+     *
+     * @param realm  The realm to query within.
+     * @param className  The type to query.
+     * @return {@link RealmQuery} object. After building the query call one of the {@code find*} methods
+     * to run it.
+     */
+    public static <E extends RealmObject> RealmQuery<E> createDynamicQuery(DynamicRealm realm, String className) {
+        return new RealmQuery<E>(realm, className);
+    }
+
+    /**
+     * Creates a sub-query from a existing {@link RealmResults}.
+     *
+     * @param queryResults   A existing @{link io.realm.RealmResults} to query against.
+     * @return {@link RealmQuery} object. After building the query call one of the {@code find*} methods
+     * to run it.
      */
-    public RealmQuery(Realm realm, Class<E> clazz) {
+
+    @SuppressWarnings("unchecked")
+    public static <E extends RealmObject> RealmQuery<E> createSubQuery(RealmResults<E> queryResults) {
+        if (queryResults.classSpec != null) {
+            return new RealmQuery<E>(queryResults, queryResults.classSpec);
+        } else {
+            return new RealmQuery(queryResults, queryResults.className);
+        }
+    }
+
+    /**
+     * Creates a sub-query from a existing {@link RealmList}.
+     *
+     * @param list   A existing @{link io.realm.RealmList} to query against.
+     * @return {@link RealmQuery} object. After building the query call one of the {@code find*} methods
+     * to run it.
+     */
+    @SuppressWarnings("unchecked")
+    public static <E extends RealmObject> RealmQuery<E> createSubQuery(RealmList<E> list) {
```

Comment:
We are still using `where` in the public API. However using `where` as a factory constructor is really `ambiguous` so I would rather not use that here.